compiler/rustc_mir_build/src/build/expr/as_constant.rs

```diff
@@ -8,33 +8,27 @@ use rustc_middle::ty::CanonicalUserTypeAnnotation;
 impl<'a, 'tcx> Builder<'a, 'tcx> {
     /// Compile `expr`, yielding a compile-time constant. Assumes that
     /// `expr` is a valid compile-time constant!
-    crate fn as_constant<M>(&mut self, expr: M) -> Constant<'tcx>
-    where
-        M: Mirror<'tcx, Output = Expr<'tcx>>,
-    {
-        let expr = self.hir.mirror(expr);
-        self.expr_as_constant(expr)
-    }
-
-    fn expr_as_constant(&mut self, expr: Expr<'tcx>) -> Constant<'tcx> {
+    crate fn as_constant(&mut self, expr: &Expr<'_, 'tcx>) -> Constant<'tcx> {
         let this = self;
         let Expr { ty, temp_lifetime: _, span, kind } = expr;
         match kind {
```

Comment:
I think you could reduce the fallout by dereferencing here

Reply:
The most problematic one here is `span`: I can't deref `expr` since `kind` is not `Copy`... but maybe it should be, since now it only holds references instead of `Box`es?

Reply:
Actually, I can write `ref kind` for now.